components/action-mgt/org.wso2.carbon.identity.action.management/src/main/java/org/wso2/carbon/identity/action/management/dao/impl/ActionManagementDAOImpl.java

```diff
@@ -415,6 +463,53 @@ private void updateAuthentication(String actionId, Authentication updatingAuthen
         updateActionPropertiesInDB(actionId, nonSecretAuthenticationProperties, tenantId);
     }
 
+    private void addRuleReference(ActionDTO actionDTO, Integer tenantId) throws ActionMgtServerException {
+
+        if (actionDTO.getActionRule() == null) {
+            return;
+        }
+
+        Map<String, String> propertiesMap =
+                Collections.singletonMap(ActionMgtConstants.RULE_PROPERTY, actionDTO.getActionRule().getId());
+        try {
+            addActionPropertiesToDB(actionDTO.getId(), propertiesMap, tenantId);
+        } catch (TransactionException e) {
+            throw new ActionMgtServerException("Error while adding the reference for the Rule in Action.", e);
+        }
+    }
+
+    private void updateRuleReference(ActionDTO updatingActionDTO, ActionDTO existingActionDTO, Integer tenantId)
+            throws ActionMgtServerException {
+
+        if (existingActionDTO.getActionRule() == null && updatingActionDTO.getActionRule() != null) {
+            // This means a new action rule is added when updating the action. Add the rule reference.
+            addRuleReference(updatingActionDTO, tenantId);
+        } else if (existingActionDTO.getActionRule() != null && updatingActionDTO.getActionRule() == null) {
+            // This means the existing action rule is removed when updating the action. Remove the rule reference.
+            deleteRuleReference(updatingActionDTO, tenantId);
+        }
+    }
+
+    private void deleteRuleReference(ActionDTO actionDTO, Integer tenantId) throws ActionMgtServerException {
+
+        try {
+            deleteActionPropertiesInDB(actionDTO.getId(),
+                    Collections.singletonList(ActionMgtConstants.RULE_PROPERTY), tenantId);
+        } catch (TransactionException e) {
+            throw new ActionMgtServerException("Error while removing the reference for the Rule in Action.", e);
+        }
+    }
+
+    private ActionRule populateRule(Map<String, String> propertiesFromDB, Integer tenantId) {
+
+        if (!propertiesFromDB.containsKey(ActionMgtConstants.RULE_PROPERTY)) {
+            return null;
```

Comment:
Shall we add a comment saying that the rule is an optional attribute

Reply:
Is that necessary. The code says and the API says that